### PR TITLE
PLUGIN-1418: add argument for BQ temp staging bucket names

### PR DESCRIPTION
**Pull request by itsankit-google**
Follow up PR of #1152.

It includes the support of creating common staging bucket for Dataplex and BQ pushdown plugins.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java [hunk 1]
```diff
@@ -858,4 +859,30 @@ public static Map<String, String> getJobTags(String jobType) {
...
+    // we must set this prefix along with the destination location in order to create/reuse the bucket.
+    // Otherwise, if temp bucket name is defined, or a prefix is not set, the configureBucket method will prepare
+    // for a new bucket creation.
+    if (Strings.isNullOrEmpty(bucket) && bucketPrefix != null) {
+      // If the destination dataset exists, use the dataset location. Otherwise, use location from configuration.
+      String datasetLocation = dataset != null ? dataset.getLocation() : configLocation;
+      // Get the bucket name for the specified location.
+      bucket = BigQueryUtil.getBucketNameForLocation(bucketPrefix, datasetLocation);
+    }
+    return bucket;
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java:887]: How does this behave is the supplied Bucket is nullable and there is no prefix set? As far as I can tell, the returned `bucket` will be null.

If this is the case, please add the `@Nullable` annotation to this function.

Following question is: Should this function return a bucket name if not supplied and no prefix is set?

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 2]
```diff
@@ -153,12 +153,14 @@ public void prepareRun(SQLEngineContext context) throws Exception {
...
-    dataset = sqlEngineConfig.getDataset();
-    bucket = sqlEngineConfig.getBucket() != null ? sqlEngineConfig.getBucket() : "bqpushdown-" + runId;
+    datasetName = sqlEngineConfig.getDataset();
 
     // Initialize BQ and GCS clients.
     bigQuery = GCPUtils.getBigQuery(project, credentials);
     storage = GCPUtils.getStorage(project, credentials);
+    Dataset dataset = bigQuery.getDataset(DatasetId.of(datasetProject, datasetName));
+    bucket = BigQueryUtil.getStagingBucketName(context.getRuntimeArguments(), sqlEngineConfig.getLocation(),
+                                               dataset, null);
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: The final argument should be `sqlEngineConfig.getBucket()`

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java [hunk 3]
```diff
@@ -169,11 +171,11 @@ public void prepareRun(SQLEngineContext context) throws Exception {
...
                                                    cmekKeyName, sqlEngineConfig.getServiceAccountType());
     // Create resources needed for this execution
-    BigQuerySinkUtils.createResources(bigQuery, storage, DatasetId.of(datasetProject, dataset), bucket,
+    BigQuerySinkUtils.createResources(bigQuery, storage, DatasetId.of(datasetProject, datasetName), bucket,
                                       sqlEngineConfig.getLocation(), cmekKeyName);
     // Configure GCS bucket that is used to stage temporary files.
-    // If the bucket is created for this run, mar it for deletion after executon is completed
-    BigQuerySinkUtils.configureBucket(configuration, bucket, runId, sqlEngineConfig.getBucket() == null);
+    // If the bucket is created for this run, mark it for deletion after executon is completed
+    BigQuerySinkUtils.configureBucket(configuration, bucket, "bqpushdown-" + runId);
```
Comment (fernst) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: nit: The second argument should be in a variable that outlines it's use

`String fallbackBucketName =  "bqpushdown-" + runId;`

**Review by fernst — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java (hunk 1 above)
Comment (itsankit-google) [reply src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java:887]: Yeah, it is expected to return `null` in that case. Added the `@Nullable` annotation.

No we don't expect this method to return a bucket name if not supplied and if no prefix is set.
`BigQuerySinkUtils#configureBucket` and `BigQuerySourceUtils#getOrCreateBucket` expect the `bucket` argument to be `nullable` and will keep the existing behaviour.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 2 above)
Comment (itsankit-google) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: ack

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java (hunk 3 above)
Comment (itsankit-google) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java]: ack

**Review by fernst — APPROVED** [review]
lgtm

**Pull request merged**